### PR TITLE
assertion gRPC errors, details and message

### PR DESCRIPTION
**Pull request by rerorero**
Hi.
I want to write a scenario to check the details of errors but there seems to be no way to do assert with gRPC error message and details.
In this PR, if a gRPC error is returned, it can be checked by asserting `message` as an [rpc/status.proto of gRPC](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto
).
What do you think about this?

### CODE REVIEW

**Comment by zoncoen** [comment]
Thank you for this work! This idea which adds assert mechanism for gRPC errors is nice. But I'd rather not use the `expect.body` field for error message and details, because it is the field for the response body.

**Comment by zoncoen** [comment]
For example, I prefer the following format to reuseing the `body` field.

```yaml
expect:
  status:
    code: InvalidArgument
    message: invalid argument
    details:
    - fieldViolations:
      - field: name
        description: "can't use '@'"
```

How would this feel?

**Comment by rerorero** [comment]
it would make sense, I'll fix it.

**Comment by rerorero** [comment]
Finally, finished @zoncoen PTAL

File: protocol/grpc/expect.go [hunk 1]
```diff
@@ -16,8 +19,112 @@ import (
...
+			d, ok := i.(interface{ String() string })
+			if ok {
+				details = append(details, fmt.Sprintf(format, name, d.String()))
+				continue
+			}
+			e, ok := i.(interface{ Error() string })
+			if ok {
+				details = append(details, fmt.Sprintf(format, name, e.Error()))
+				continue
+			}
```
Comment (zoncoen) [inline protocol/grpc/expect.go]: We should add some message if `i` doesn't implement not only `String()` but also `Error()`.

File: protocol/grpc/expect.go [hunk 2]
```diff
@@ -16,8 +19,112 @@ import (
...
+		if len(expectDetailMap) != 1 {
+			return errors.Errorf("invalid expect status.details[%d]:"+
+				"An element of status.details list must be a map of size 1 with the detail message name as the key and the value as the detail message object.", i)
+		}
+		expectName, ok := expectDetailMap[0].Key.(string)
+		if !ok {
+			return errors.Errorf("invalid expect status.details[%d]: detail name must be a string", i)
+		}
+		expectDetail := expectDetailMap[0].Value
+		template, err := ctx.ExecuteTemplate(expectDetail)
```
Comment (zoncoen) [inline protocol/grpc/expect.go]: Let's do the same here please.

File: protocol/grpc/expect.go [hunk 3]
```diff
@@ -43,68 +156,69 @@ func (e *Expect) Build(ctx *context.Context) (assert.Assertion, error) {
...
+	if got, expected := strconv.Itoa(int(int32(stErr.Code))), expectedCode; got == expected {
+		return nil
 	}
 
-	return errors.Errorf(`expected code is "%s" but got "%s": %s: %s`, expectedCode, stErr.Code().String(), err, strings.Join(details, ", "))
+	return errors.Errorf(`expected code is "%s" but got "%s": %s: %s`, expectedCode, stErr.Code.String(), stErr.Message, stErr.detailsString())
 }
 
-func extract(v interface{}) (proto.Message, error, error) {
+func extract(v interface{}) (interface{}, simpleStatusError, error) {
```
[the remark this answers is not in the document]
Comment (zoncoen) [reply protocol/grpc/expect.go]: I think `simpleStatusError` is no need. Would you add `(*Expect).assertError()` which converts to `status.Status` and assert the code, message and details?

[the remark this answers is not in the document]
Comment (rerorero) [reply protocol/grpc/expect.go]: sure!

> Why you change the return value proto.Message to interface{}?

Oops, I forgot to revert it. I changed it at the first commit where I put status error into `message`.

**Comment by rerorero** [comment]
@zoncoen corrected the problem you pointed out, PTAL!

File: protocol/grpc/expect.go (hunk 1 above)
Comment (rerorero) [reply protocol/grpc/expect.go]: fixed here https://github.com/zoncoen/scenarigo/pull/12/files#diff-230f4aad0c1511967cd1ed69f93e274bR163

File: protocol/grpc/expect.go [hunk 4]
```diff
@@ -16,8 +19,112 @@ import (
...
+		})
+	}
+	return detailsMaps
+}
+
+func (e *ExpectStatus) assertMessage(ctx *context.Context, actualMessage string) error {
+	if e.Message == "" {
+		return nil
+	}
+	expectedMessage, err := ctx.ExecuteTemplate(e.Message)
```
[the remark this answers is not in the document]
Comment (rerorero) [reply protocol/grpc/expect.go]: removed all ExecuteTemplate in this PR

File: protocol/grpc/expect.go [hunk 5]
```diff
@@ -16,8 +19,112 @@ import (
...
+	Details []yaml.MapSlice `yaml:"details"`
+}
+
+// simpleStatusError is used for status.Status whose details converted into proto.Message
+// to be comparable with simple key-map yaml
+type simpleStatusError struct {
+	Code    codes.Code
+	Message string
+	Details []interface{}
+}
```
[the remark this answers is not in the document]
Comment (rerorero) [reply protocol/grpc/expect.go]: `simpleStatusError` was removed

File: protocol/grpc/expect.go [hunk 6]
```diff
@@ -33,7 +43,13 @@ func (e *Expect) Build(ctx *context.Context) (assert.Assertion, error) {
...
 			return err
 		}
-		if err := e.assertCode(callErr); err != nil {
+		if err := e.assertStatusCode(callErr); err != nil {
+			return err
+		}
+		if err := e.assertStatusMessage(callErr); err != nil {
+			return err
+		}
+		if err := e.assertStatusDetails(callErr); err != nil {
```
[the remark this answers is not in the document]
Comment (zoncoen) [reply protocol/grpc/expect.go]: Convert `error` to `status.Status` before calling `assertStatusXXX` functions, please.

[the remark this answers is not in the document]
Comment (rerorero) [reply protocol/grpc/expect.go]: okay :)

[the remark this answers is not in the document]
Comment (rerorero) [reply protocol/grpc/expect.go]: fixed!

**Pull request merged**